v1/brokers/amqp.go

```diff
@@ -170,25 +175,31 @@ func (b *AMQPBroker) consume(deliveries <-chan amqp.Delivery, concurrency int, t
 		case err := <-errorsChan:
 			return err
 		case d := <-deliveries:
-			if concurrency > 0 {
-				// get worker from pool (blocks until one is available)
-				<-pool
+			// Geting worker from pool (blocks until one is available)
+			// This shouldn't return any errors since it fails only if the context is
+			// canceled/contains error but just in case we handle failure anyway
+			if err := pool.Acquire(context.Background(), 1); err != nil {
+				return err
 			}
 
 			wg.Add(1)
 
 			// Consume the task inside a gotourine so multiple tasks
 			// can be processed concurrently
 			go func() {
-				defer wg.Done()
-
-				if err := b.consumeOne(d, taskProcessor); err != nil {
-					errorsChan <- err
-				}
-
-				if concurrency > 0 {
-					// give worker back to pool
-					pool <- struct{}{}
+				err := b.consumeOne(d, taskProcessor)
+
+				wg.Done()
+				// give worker back to pool
+				pool.Release(1)
+
+				if err != nil {
+					select {
+					case <-quitChan:
+						// main loop exited; ignoring error
+					case errorsChan <- err:
```

Comment:
please use `default:` for consistency's sake

Reply:
A safer (for the deadlock) solution here would be setting the capacity of the `errorsChan` to match the semaphore's weight, so you can do like this (and you're guaranteed not to have hanging goroutines (can still happen here)):
```
errorsChan := make(chan error, int64(concurrency))
...
// Consume the task inside a gotourine so multiple tasks
// can be processed concurrently
go func() {
	defer wg.Done()
	defer pool.Release(1)

	err := b.consumeOne(d, taskProcessor)
	if err != nil {
		errorsChan <- err
	}
}()
```
This way you can also save completely the usage of the `quitChan`

Reply:
why the `default` case?

Reply:
```
if err != nil {
	select {
	case <-quitChan:
	default:
		errorsChan <- err
	}
}
```
it's a more logical way of putting it - it should normally pass the error upstream, unless the quitChan is closed, for which there's a special case.

Reply:
`wg.Done` and `pool.Release(1)` must be called before `errorsChan <- err` to prevent deadlock. `defer` has no advantages here and introduces the deadlock we are trying to avoid

Reply:
I'm aware of that, and that's why setting the channel's capacity is a solution to the deadlock for the deferred release/done that also allows you to not introduce the `quitChan`.

Reply:
I considered using a buffered channel but this approach over-allocates memory though (in case of `concurrency` of `0` it over-allocates by A LOT) as we're only interested one error at most

Reply:
See my last sentence about the pool size ;)